File: solutions/python/dp-1d/perfect-squares.py
# LeetCode 279 — Perfect Squares (Medium)
# Category: 1-D Dynamic Programming · Approach: DP
# Time: O(n·√n) | Space: O(n)
# Source: https://leetcode.com/problems/perfect-squares/

def num_squares(n: int) -> int:
    dp = [0] + [float("inf")] * n
    for x in range(1, n + 1):
        s = 1
        while s * s <= x:
            dp[x] = min(dp[x], dp[x - s * s] + 1)
            s += 1
    return int(dp[n])

File: solutions/typescript/dp-1d/perfect-squares.ts
// LeetCode 279 — Perfect Squares (Medium)
// Category: 1-D Dynamic Programming · Approach: DP
// Time: O(n·√n) | Space: O(n)
// Source: https://leetcode.com/problems/perfect-squares/

function numSquares(n: number): number {
  const dp = new Array<number>(n + 1).fill(Infinity);
  dp[0] = 0;
  for (let x = 1; x <= n; x++)
    for (let s = 1; s * s <= x; s++)
      dp[x] = Math.min(dp[x], dp[x - s * s] + 1);
  return dp[n];
}
